source/secure_tunneling.c

```diff
@@ -473,198 +396,1395 @@ bool secure_tunneling_send_data_call(struct aws_websocket *websocket, struct aws
+    if (next_state == secure_tunnel->current_state) {
+        return;
+    }
+
+    AWS_LOGF_DEBUG(
+        AWS_LS_IOTDEVICE_SECURE_TUNNELING,
+        "id=%p: switching current state from %s to %s",
+        (void *)secure_tunnel,
+        aws_secure_tunnel_state_to_c_string(secure_tunnel->current_state),
+        aws_secure_tunnel_state_to_c_string(next_state));
+
+    switch (next_state) {
+        case AWS_STS_STOPPED:
+            s_change_current_state_to_stopped(secure_tunnel);
+            break;
+        case AWS_STS_CONNECTING:
+            s_change_current_state_to_connecting(secure_tunnel);
+            break;
+        case AWS_STS_CONNECTED:
+            s_change_current_state_to_connected(secure_tunnel);
+            break;
+        case AWS_STS_CLEAN_DISCONNECT:
+            s_change_current_state_to_clean_disconnect(secure_tunnel);
+            break;
+        case AWS_STS_WEBSOCKET_SHUTDOWN:
+            s_change_current_state_to_websocket_shutdown(secure_tunnel);
+            break;
+        case AWS_STS_PENDING_RECONNECT:
+            s_change_current_state_to_pending_reconnect(secure_tunnel);
+            break;
+        case AWS_STS_TERMINATED:
+            s_change_current_state_to_terminated(secure_tunnel);
+            return;
+    }
+
+    s_reevaluate_service_task(secure_tunnel);
+}
+
+static bool s_is_valid_desired_state(enum aws_secure_tunnel_state desired_state) {
+    switch (desired_state) {
+        case AWS_STS_STOPPED:
+        case AWS_STS_CONNECTED:
+        case AWS_STS_TERMINATED:
+            return true;
+        default:
+            return false;
+    }
+}
+
+struct aws_secure_tunnel_change_desired_state_task {
+    struct aws_task task;
+    struct aws_allocator *allocator;
+    struct aws_secure_tunnel *secure_tunnel;
+    enum aws_secure_tunnel_state desired_state;
+};
+
+static void s_change_state_task_fn(struct aws_task *task, void *arg, enum aws_task_status status) {
+    (void)task;
+
+    struct aws_secure_tunnel_change_desired_state_task *change_state_task = arg;
+    struct aws_secure_tunnel *secure_tunnel = change_state_task->secure_tunnel;
+    enum aws_secure_tunnel_state desired_state = change_state_task->desired_state;
+    if (status != AWS_TASK_STATUS_RUN_READY) {
+        goto done;
+    }
+
+    if (secure_tunnel->desired_state != desired_state) {
+        AWS_LOGF_INFO(
+            AWS_LS_IOTDEVICE_SECURE_TUNNELING,
+            "id=%p: changing desired secure_tunnel state from %s to %s",
+            (void *)secure_tunnel,
+            aws_secure_tunnel_state_to_c_string(secure_tunnel->desired_state),
+            aws_secure_tunnel_state_to_c_string(desired_state));
+
+        secure_tunnel->desired_state = desired_state;
+
+        s_reevaluate_service_task(secure_tunnel);
+    }
+
+done:
+
+    if (desired_state != AWS_STS_TERMINATED) {
+        aws_secure_tunnel_release(secure_tunnel);
+    }
+
+    aws_mem_release(change_state_task->allocator, change_state_task);
```

Comment:
I think it's impossible for it to be null because there's a null check on the task before the task is created that calls this function. If the task is null, maybe we want it to crash because how the hell did that happen? =P